_posts/2023-07-29-oauth-authentication-scala-http4s.md

```diff
@@ -0,0 +1,672 @@
+import com.xonal.config.{Config, configuration}
+import org.http4s.implicits.uri
+import org.http4s.headers.{Accept, Authorization}
+import cats.syntax.all.*
+
+object OauthImpl {
+  final case class GithubResponse(
+      accessToken: String,
+      tokenType: String,
+      scope: String
+  )
+
+  private object GithubResponse{
+    given decoder: Decoder[GithubResponse] = Decoder.instance { h =>
+      for
+        access_token <- h.get[String]("access_token")
+        token_type <- h.get[String]("token_type")
+        scope <- h.get[String]("scope")
+      yield GithubResponse(access_token, token_type, scope)
+    }
+  }
+
+  private def decodeJson(jsonString: String): Either[Error, GithubResponse] =
+    decode[GithubResponse](jsonString)
+
+  private def getJsonString[F[_]: Async](req: Request[F]): F[String] =
+    EmberClientBuilder
+      .default[F]
+      .build
+      .use(client => client.expect[String](req))
+
+  private def fetchJsonString[F[_]: Async](
+      code: String,
+      config: Config
+  ): F[String] = {
+    val form = UrlForm(
+      "client_id" -> config.api.key,
+      "client_secret" -> config.api.secret.value,
+      "code" -> code
+    )
+
+    val req = Request[F](
+      Method.POST,
+      uri"https://github.com/login/oauth/access_token",
+      headers = Headers(Accept(MediaType.application.json))
+    ).withEntity(form)
+
+    getJsonString(req)
+  }
+
+  private def fetchGithubDetails[F[_]: Async](
+      access_token: String
+  ): F[String] = {
+    val req = Request[F](
+      Method.GET,
+      uri"https://api.github.com/user/emails",
+      headers = Headers(
+        Accept(MediaType.application.json),
+        Authorization(Credentials.Token(AuthScheme.Bearer, access_token))
+      )
+    )
+    getJsonString(req)
+  }
+
+  def getOauthResults[F[_]: Async](code: String): F[String] =
+    for
+      config <- configuration.load[F]
+      decodedJson <- fetchJsonString[F](code, config).map(decodeJson(_))
+      githubDetails <- decodedJson match
+        case Right(v)  => fetchGithubDetails(v.accessToken)
+        case Left(err) => err.pure[F].map(_.getMessage)
+    yield githubDetails
+
+}
+```
+
+### 2.5. The application routes.
+
+The next step is to define routes for our application. Let's create `AppRoutes.scala` in the following path, `oauth/src/main/scala/com/xonal/routes/AppRoutes.scala`.
+In this section we will define two routes, the first will show our landing page which will be static HTML and the second will handle the call back from GitHub after the user authorizes our Oauth application.
+
+```scala
+package com.xonal.routes
+
+import org.http4s.*
+import org.http4s.dsl.Http4sDsl
+import cats.effect.Async
+import fs2.io.file.Path
+
+object GithubRoutes{
+  def githubRoutes[F[_]: Async]: HttpRoutes[F] = {
+    val dsl = Http4sDsl[F]
+    import dsl.{Path as _, *}
+    HttpRoutes.of[F] {
+      case request @ GET -> Root / "index.html" =>
+        StaticFile
+          .fromPath(
```

Comment:
can use .fromString and remove the weird path import

Reply:
Updated